cni/network/invoker_cns.go

```diff
@@ -70,94 +72,143 @@ func (invoker *CNSIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, erro
+		}
 	}
 
-	info := IPv4ResultInfo{
-		podIPAddress:       response.PodIpInfo.PodIPConfig.IPAddress,
-		ncSubnetPrefix:     response.PodIpInfo.NetworkContainerPrimaryIPConfig.IPSubnet.PrefixLength,
-		ncPrimaryIP:        response.PodIpInfo.NetworkContainerPrimaryIPConfig.IPSubnet.IPAddress,
-		ncGatewayIPAddress: response.PodIpInfo.NetworkContainerPrimaryIPConfig.GatewayIPAddress,
-		hostSubnet:         response.PodIpInfo.HostPrimaryIPInfo.Subnet,
-		hostPrimaryIP:      response.PodIpInfo.HostPrimaryIPInfo.PrimaryIP,
-		hostGateway:        response.PodIpInfo.HostPrimaryIPInfo.Gateway,
-	}
+	addResult := IPAMAddResult{}
 
-	// set the NC Primary IP in options
-	addConfig.options[network.SNATIPKey] = info.ncPrimaryIP
+	for i := 0; i < len(response.PodIPInfo); i++ {
+		info := IPResultInfo{
+			podIPAddress:       response.PodIPInfo[i].PodIPConfig.IPAddress,
+			ncSubnetPrefix:     response.PodIPInfo[i].NetworkContainerPrimaryIPConfig.IPSubnet.PrefixLength,
+			ncPrimaryIP:        response.PodIPInfo[i].NetworkContainerPrimaryIPConfig.IPSubnet.IPAddress,
+			ncGatewayIPAddress: response.PodIPInfo[i].NetworkContainerPrimaryIPConfig.GatewayIPAddress,
+			hostSubnet:         response.PodIPInfo[i].HostPrimaryIPInfo.Subnet,
+			hostPrimaryIP:      response.PodIPInfo[i].HostPrimaryIPInfo.PrimaryIP,
+			hostGateway:        response.PodIPInfo[i].HostPrimaryIPInfo.Gateway,
+		}
 
-	log.Printf("[cni-invoker-cns] Received info %+v for pod %v", info, podInfo)
+		// set the NC Primary IP in options
+		addConfig.options[network.SNATIPKey] = info.ncPrimaryIP
 
-	// set result ipconfigArgument from CNS Response Body
-	ip, ncipnet, err := net.ParseCIDR(info.podIPAddress + "/" + fmt.Sprint(info.ncSubnetPrefix))
-	if ip == nil {
-		return IPAMAddResult{}, errors.Wrap(err, "Unable to parse IP from response: "+info.podIPAddress+" with err %w")
-	}
+		log.Printf("[cni-invoker-cns] Received info %+v for pod %v", info, podInfo)
 
-	ncgw := net.ParseIP(info.ncGatewayIPAddress)
-	if ncgw == nil {
-		if invoker.ipamMode != util.V4Overlay {
-			return IPAMAddResult{}, errors.Wrap(errInvalidArgs, "%w: Gateway address "+info.ncGatewayIPAddress+" from response is invalid")
+		// set result ipconfigArgument from CNS Response Body
+		ip, ncIPNet, err := net.ParseCIDR(info.podIPAddress + "/" + fmt.Sprint(info.ncSubnetPrefix))
+		if ip == nil {
+			return IPAMAddResult{}, errors.Wrap(err, "Unable to parse IP from response: "+info.podIPAddress+" with err %w")
 		}
 
-		ncgw, err = getOverlayGateway(ncipnet)
-		if err != nil {
-			return IPAMAddResult{}, err
+		ncgw := net.ParseIP(info.ncGatewayIPAddress)
+		if ncgw == nil {
+			if (invoker.ipamMode != util.V4Overlay) && (invoker.ipamMode != util.DualStackOverlay) {
+				return IPAMAddResult{}, errors.Wrap(errInvalidArgs, "%w: Gateway address "+info.ncGatewayIPAddress+" from response is invalid")
+			}
+
+			if net.ParseIP(info.podIPAddress).To4() != nil {
+				ncgw, err = getOverlayGateway(ncIPNet)
```

Comment:
@paulyufan we get the GW from the pod subnet in case of ipv4. We are using hardcoded value for ipv6. Why is that?

Reply:
Hi Ashvin, there are two reasons:
(1) We get this default ipv6 gateway "fe80::1234:5678:9abc" from Jearyn at the beginning and we are using this as default ipv6 gateway.
(2) Jaeryn made changes recently for this reason:
```
Fixing overlay gateway bug that prevented cloud-node-manager-windows from initializing node because it couldn't reach IMDS. An additional VFP rule was implicitly plumbed by HNS to add 169.254.1.1/16 to the exception list. This is most likely based off the RouteConfiguration in the "azure" HNS network.  This additional VFP rule was skipping SNAT for any IP address in this range 169.254.0.0-169.254.255.255. We'll set the first .1 IP of podcidr as the gateway now.
```
Do we need to use this to reach IMDS?